docs/CustomAuthenticationProvider.md

```diff
@@ -0,0 +1,39 @@
+# Using Custom Authentication Provider
+
+Using preferred choice of Authentication library for authenticating with Microsoft is possible.
+
+## Step by step procedure
+
+### Implement AuthenticationProvider
+
+Create own implementation of Authentication provider which implements [AuthenticationProvider](../src/iAuthenticationProvider.ts) interface.
+
+```typescript
+// MyAuthenticationProvider.ts
+import { AuthenticationProvider } from "@microsoft/microsoft-graph-client";
+
+class MyAuthenticationProvider implements AuthenticationProvider {
+
+    /**
+     * This method will get called before every request to the msgraph server
+     * This should return a Promise that resolves to an accessToken (in case of success) or rejects with error (in case of failure)
+     * Basically this method will contain the implementation for getting and refreshing accessTokens
+     */
+    public async getAccessToken(): Promise<any> {
```

Comment:
I'm not sure this interface is flexible enough.  I think we need to pass an array of scopes to enable incremental consent and pass some kind of user identifier in case we are in a multi-user application.

Reply:
The example you have seen is a sample implementation of CustomAuthenticationProvider, we have an option to pass an array of scopes to for incremental consent in [MSALAuthenticationProvider](https://github.com/microsoftgraph/msgraph-sdk-javascript/blob/Middleware/src/MSALAuthenticationProvider.ts) with `.addScopes()` method.
For multi user application also we can add similar kind of method to switch the user.

Reply:
Ok. Never mind.  There is a default set of scopes that are passed in the constructor.  There is no harm in having the ability to add to that default set of scopes.  If we need to allow the ability to request a token for a smaller set of scopes in the future we can add an optional parameter to getAccessToken that accepts a new array of scopes.